keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1194 @@
+ability to easily migrate the workload pods from one node to another. This also applies to
+applications with larger number of replicas that prefer to surge (upscale) pods first instead of
+downscaling.
+
+### Notes/Constraints/Caveats (Optional)
+
+DaemonSet pods and mirror pods can be evacuated, but cannot be evicted because they can run
+critical services.
+
+### Risks and Mitigations
+
+#### TBD: Security
+
+#### Disruptive Eviction
+
+When using kubectl drain, pods without owning controller and pods with local storage
+(having `EmptyDir` volumes) are not evicted by default. We have decided to evict most
+of the pods (except DaemonSet and mirror pods) by default. In the motivation section of the
+[Declarative Node Maintenance KEP](https://github.com/kubernetes/enhancements/pull/4213),
+we can see many administrators override these default settings and many components evict all pods
+indiscriminately. There are also many ways that users use to prevent the eviction;
+PodDisruptionBudgets, validating admission webhooks, or just plain HA. Users who want to protect
+their applications in today's clusters should already be aware that they should be able to handle
+sudden evictions. Therefore, it should be okay for the evacuation controller to evict these pods.
+
+To mitigate the sudden eviction problem, users should use PodDisruptionBudgets or HA.
+
+## Design Details
+
+### Evacuation
+
+We will introduce a new type called `evacuation.coordination.k8s.io`  to enforce the contract
+between the evacuation instigators and the evacuators. This type is a bit similar to
+`leases.coordination.k8s.io` in that it requires multiple actors to synchronize the state. Which in
+our case is the progress of the evacuation.
+
+### Evacuation Instigator
+[Evacuation Instigator](#evacuation-instigator) section provides a general overview.
+
+There can be many evacuation instigators for a single Evacuation. 
+
+When an instigator decides that a pod needs to be evacuated, it should create an Evacuation:
+- With a predictable name in the following format: `${POD_UID}-${POD_NAME_PREFIX}`. The pod name
+  might be only partial as the full eviction name is limited to 253 characters.
+- `.spec.podRef` should be set to fully identify the pod. Similar to the name, the UID should be
+  specified to ensure that we do not evacuate a pod with the same name that appears immediately
+  after the previous one is removed.
+- `.spec.acceptDeadlineSeconds` should be set to a reasonable value. For example, 600 (10m) to
+  allow reasonably fast feedback, but still give the potential evacuator sufficient time to begin
+  the evacuation.
+- `.spec.progressDeadlineSeconds` should be set to a reasonable value. For example, 3600 (1h) to
+  allow for potential evacuator disruptions.
+- The pod labels should be copied to the Evacuation labels to allow for custom label selectors when
+  observing the evacuations.
```

Comment:
We should clearly explain in this KEP why the Evacuation can't be a subresource.

Reply:
I do mean that.

Reply:
Ack, added this to alternatives.